dask/array/core.py

```diff
@@ -1938,14 +1997,85 @@ def normalize_chunks(chunks, shape=None):
             raise ValueError("Input array has %d dimensions but the supplied "
                              "chunks has only %d dimensions" %
                              (len(shape), len(chunks)))
-        if not all(c == s or (math.isnan(c) and math.isnan(s))
+        if not all(c == s or (math.isnan(c) or math.isnan(s))
                    for c, s in zip(map(sum, chunks), shape)):
             raise ValueError("Chunks do not add up to shape. "
                              "Got chunks=%s, shape=%s" % (chunks, shape))
 
     return tuple(tuple(int(x) if not math.isnan(x) else x for x in c) for c in chunks)
 
 
+def auto_chunks(chunks, shape, limit, dtype, previous_chunks=None):
+    """ Determine automatic chunks
+
+    See also
+    --------
+    normalize_chunks: for full docstring and parameters
+    """
+    if limit is None:
+        limit = config.get('array.chunk-size')
+        if isinstance(limit, str):
+            limit = parse_bytes(limit)
+
+    if dtype is None:
+        raise TypeError("DType must be known for auto-chunking")
+    if dtype.hasobject:
+        raise NotImplementedError("Can not use auto rechunking with object dtype")
+    limit = max(1, limit // dtype.itemsize)
+    chunks = list(chunks)
+
+    autos = {i for i, c in enumerate(chunks) if c == 'auto'}
+    if not autos:
```

Comment:
Can we end up here without "auto"s? In any case, this should come before the check for object, since if not actually changing anything, should be allowed to not-change for object type too.

Reply:
We can end up here in the recursive case.  I've moved this block to the top of the function.